File: leetcode.com/0513_find_bottom_left_tree_value/solution.go
/**
* Definition for a binary tree node.
* type TreeNode struct {
*     Val int
*     Left *TreeNode
*     Right *TreeNode
* }
*/

var r_depth int
var r_val int

func dfs(node *TreeNode, depth int, left bool) {
	if node == nil {
		return
	}

	if left && depth > r_depth {
		r_val = node.Val
		r_depth = depth
	}

	dfs(node.Left, depth + 1, true)
	dfs(node.Right, depth + 1, node.Left == nil)
}

func findBottomLeftValue(root *TreeNode) int {
	r_val = root.Val
	r_depth = 0
	dfs(root, 0, true)
	return r_val;
}

File: leetcode.com/0513_find_bottom_left_tree_value/solution.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

int r_depth;
int r_val;

void dfs(struct TreeNode* node, int depth, bool left)
{
    if (!node)
        return;

    if (left && depth > r_depth) {
        r_val = node->val;
        r_depth = depth;
    }

    dfs(node->left, depth + 1, true);
    dfs(node->right, depth + 1, false || (node->left == NULL));
}

int findBottomLeftValue(struct TreeNode* root)
{
    r_val = root->val;
    r_depth = 0;

    dfs(root, 0, true);
    return r_val;
}
